Prometheus/Histogram.cs

```diff
@@ -83,36 +89,47 @@ internal Child(Histogram parent, LabelSequence instanceLabels, LabelSequence fla
                     CanonicalLabel.Empty,
                     cancel,
                     _sum.Value,
+                    ObservedExemplar.Empty,
                     suffix: SumSuffix);
                 await serializer.WriteMetricPointAsync(
                     Parent.NameBytes,
                     FlattenedLabelsBytes,
                     CanonicalLabel.Empty,
                     cancel,
                     _bucketCounts.Sum(b => b.Value),
+                    ObservedExemplar.Empty,
                     suffix: CountSuffix);
 
                 var cumulativeCount = 0L;
 
                 for (var i = 0; i < _bucketCounts.Length; i++)
                 {
+                    ObservedExemplar cp;
+                    lock (_exemplars)
```

Comment:
RFC

Reply:
It looks correct, yeah - it performs a shallow copy and that's it.

However, I am not very keen on having locks in the hot path (`ObserveInternal()`) - the metric objects try to be very lock-free in general because things like HTTP servers can funnel a lot of requests (handled by many threads) into observations on the same metric instances. So far I have managed to keep all except for Summary lock-free and would ideally like to keep it so.

You can see the potential cost of multithreaded access in `MeasurementBenchmarks` - already without locks it is rather slow (though artificially slow as there is no "other work" done in the benchmark). I am afraid that adding locks might create even more potential slowdown.

Example of differences on my PC (True/False column indicating use of exemplars; I will commit my exemplar-enhanced benchmark code shortly):

```
| MeasurementPerformance |           100000 |          16 |          Counter |         False | 124,351.9 us | 2,163.73 us | 2,023.96 us | 123,847.7 us |                    - |          13.0000 |     384 B |
| MeasurementPerformance |           100000 |          16 |          Counter |          True | 230,592.2 us | 3,085.85 us | 2,886.51 us | 229,846.1 us |                    - |        6124.0000 |     384 B |
|
| MeasurementPerformance |           100000 |          16 |        Histogram |         False | 212,602.3 us | 1,505.49 us | 1,408.24 us | 212,393.8 us |                    - |          14.0000 |     384 B |
| MeasurementPerformance |           100000 |          16 |        Histogram |          True | 242,188.5 us | 2,762.03 us | 2,583.61 us | 241,092.3 us |                    - |        1445.0000 |     384 B |
|
```

The total duration here, of course, also includes `ObservedExemplar.Update()` and is not only the lock contention overhead. The 2nd column from the right is the lock contention count.

A lock-free approach would be preferable here to try ensure that capturing data is cheap. How to do that, though? What if we turn it into a reference type and just pool them to avoid the constant allocation overhead?

* [ObjectPool](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.objectpool.objectpool.create?view=dotnet-plat-ext-7.0#microsoft-extensions-objectpool-objectpool-create-1(microsoft-extensions-objectpool-ipooledobjectpolicy((-0)))) can manage a set of `ObservedExemplar` instances.
* In `ObserveInternal()` we get a new instance from the pool and use a loop over [Interlocked.CompareExchange](https://learn.microsoft.com/en-us/dotnet/api/system.threading.interlocked.compareexchange?view=net-7.0#system-threading-interlocked-compareexchange-1(-0@-0-0)) to replace it in `_exemplars`, returning the old one to the pool.
* In `CollectAndSerializeImplAsync()` we have a bit of a roadblock that we need to prevent our instance from being returned to the pool while we are still using it. However, we can perhaps do a mirrored operation, take an instance from the pool, zero it, and swap it out with the one in `_exemplars`. After serializing the exemplar, we can swap it back (if the one in `_exemplars` is still our empty instance) and return the empty instance to the pool.
* This leaves a small window where the exemplar is left empty, so in a theoretical case two concurrent metrics exports could end up with one not seeing the exemplar. But this seems like a totally OK price to pay for lock-free code - very theoretical situation, and harmless even if it happens.

What do you think? Just off the top of my head, it could be there is a big plot hole here somewhere.